scripts: runners: openocd: Enable thread awareness

nandojve opened the pull request:
Zephyr thread awareness is available for openocd but boards don't have debuggers configuration.  Add `openocd.zephyr.cfg` file to allow enable Zephyr specific configurations.  This is loaded by openocd runner automatically to complete configuration.

User still require enable CONFIG_DEBUG_THREAD_INFO=y to visualize thread debug information.

Enhancement #37205

Signed-off-by: Gerson Fernando Budke <nandojve@gmail.com>

nandojve commented:
Just a prove of concept. Not sure if this is best way but can solve for all boards.

mbolivar-nordic commented on scripts/west_commands/runners/openocd.py:
```diff
@@ -118,7 +122,8 @@ def do_run(self, command, **kwargs):
         if self.openocd_config is not None:
-            self.cfg_cmd = ['-f', self.openocd_config]
+            self.cfg_cmd = ['-f', self.openocd_config,
+                            '-f', 'openocd.zephyr.cfg']
```
Can we make this work like the jlink runner, where it's only passed to openocd if CONFIG_DEBUG_THREAD_INFO=y? You can search in the jlink.py source code for `self.thread_info_enabled` to see how it's done.

mbolivar-nordic commented on scripts/west_commands/runners/openocd.py:
```diff
@@ -41,9 +41,13 @@ def __init__(self, cfg, pre_init=None, pre_load=None,
         else:
             search_args = []
 
+        common = path.split(path.split(cfg.board_dir)[0])
+        search_args += ['-s', path.join(common[0], 'common')]
```
Is it possible to do this by using `-c` on the openocd command line, like we do the other commands, instead of using a file? If not, I think it might be worth adding `boards/common` to the `config: openocd_search:` path in build/zephyr/runners.yaml. You can do that in cmake/flash/CMakeLists.txt.

The goal is to make this work for out of tree boards, too.

mbolivar-nordic replied:
I actually think the `openocd_search` approach might be the best overall. Curious what others think.

nandojve replied:
Hi @mbolivar-nordic ,

> Is it possible to do this by using -c on the openocd command line, like we do the other commands, instead of using a file?

I gave some tries without success. I'll ask on the openocd forum.

mbolivar-nordic replied:
>  I'll ask on the openocd forum.

I think the `openocd_search` approach is probably better, so no worries here!

mbolivar-nordic reviewed (commented):
Thanks! I like the approach, just a couple of minor questions.

nandojve commented:
Hi @mbolivar-nordic ,

For some reason `self.check_output([self.openocd_cmd[0], '--version']).decode()` it is not working with OpenOCD. Because of that, I can't determine OpenOCD version. I added manually the current output to wrote code. For now, just ignore `read_version` method.

Any idea about `check_output`?

mbolivar-nordic commented:
> it is not working

Some details would help :)

mbolivar-nordic commented:
@nandojve looks like openocd is printing to stderr; try this instead:

```py
self.check_output([self.openocd_cmd[0], '--version'], stderr=subprocess.STDOUT).decode()
```

mbolivar-nordic reviewed (changes requested):
Thanks for the update!

I tested this out and found it unfortunately doesn't play well with boards/common/openocd-nrf5.board.cmake, which has its own logic for this. I think some unification is needed to make sure this works properly for all boards that support this runner.

I also don't think it's safe in general to do a whitespace split (`cfg.openocd_search.split()`) to get the search path. I understand that `RunnerConfig.openocd_search` is an `Optional[str]` right now, but I think it needs to be made into a `List[str]` instead, with something like this:

```patch
diff --git a/cmake/flash/CMakeLists.txt b/cmake/flash/CMakeLists.txt
index 786555f582..3e47ad8ea3 100644
--- a/cmake/flash/CMakeLists.txt
+++ b/cmake/flash/CMakeLists.txt
@@ -53,9 +53,11 @@ function(runners_yaml_append_config)
   endif()
   if(OPENOCD)
     runners_yaml_append("  openocd: ${OPENOCD}")
-  endif()
-  if(OPENOCD_DEFAULT_PATH)
-    runners_yaml_append("  openocd_search: ${OPENOCD_DEFAULT_PATH} ${ZEPHYR_BASE}/boards/common")
+    runners_yaml_append("  openocd_search:")
+    runners_yaml_append("    - ${ZEPHYR_BASE}/boards/common")
+    if(OPENOCD_DEFAULT_PATH)
+      runners_yaml_append("    - ${OPENOCD_DEFAULT_PATH}")
+    endif()
   endif()
   runners_yaml_append("")
 endfunction()
diff --git a/scripts/west_commands/run_common.py b/scripts/west_commands/run_common.py
index f40c64f804..e85b7145d0 100644
--- a/scripts/west_commands/run_common.py
+++ b/scripts/west_commands/run_common.py
@@ -356,8 +356,8 @@ def get_runner_config(build_dir, yaml_path, runners_yaml, args=None):
 
         return None
 
-    def config(attr):
-        return getattr(args, attr, None) or yaml_config.get(attr)
+    def config(attr, default=None):
+        return getattr(args, attr, None) or yaml_config.get(attr, default)
 
     return RunnerConfig(build_dir,
                         yaml_config['board_dir'],
@@ -366,7 +366,7 @@ def get_runner_config(build_dir, yaml_path, runners_yaml, args=None):
                         output_file('bin'),
                         config('gdb'),
                         config('openocd'),
-                        config('openocd_search'))
+                        config('openocd_search', []))
 
 def dump_traceback():
     # Save the current exception to a file and return its path.
diff --git a/scripts/west_commands/runners/core.py b/scripts/west_commands/runners/core.py
index b460c54f60..4df9ff000b 100644
--- a/scripts/west_commands/runners/core.py
+++ b/scripts/west_commands/runners/core.py
@@ -260,7 +260,7 @@ class RunnerConfig(NamedTuple):
     bin_file: Optional[str]     # zephyr.bin path, or None
     gdb: Optional[str] = None   # path to a usable gdb
     openocd: Optional[str] = None  # path to a usable openocd
-    openocd_search: Optional[str] = None  # add this to openocd search path
+    openocd_search: List[str] = []  # add these paths to the openocd search path
 
 
 _YN_CHOICES = ['Y', 'y', 'N', 'n', 'yes', 'no', 'YES', 'NO']
diff --git a/scripts/west_commands/runners/openocd.py b/scripts/west_commands/runners/openocd.py
index 3dd06e8a42..6e68cf1e94 100644
--- a/scripts/west_commands/runners/openocd.py
+++ b/scripts/west_commands/runners/openocd.py
@@ -44,9 +44,8 @@ class OpenOcdBinaryRunner(ZephyrBinaryRunner):
         else:
             search_args = []
 
-        if cfg.openocd_search is not None:
-            for p in cfg.openocd_search.split():
-                search_args.extend(['-s', p])
+        for p in cfg.openocd_search:
+            search_args.extend(['-s', p])
         self.openocd_cmd = [cfg.openocd] + search_args
         # openocd doesn't cope with Windows path names, so convert
         # them to POSIX style just to be sure.
diff --git a/scripts/west_commands/tests/conftest.py b/scripts/west_commands/tests/conftest.py
index 139f213f86..a566e76573 100644
--- a/scripts/west_commands/tests/conftest.py
+++ b/scripts/west_commands/tests/conftest.py
@@ -15,7 +15,7 @@ RC_KERNEL_HEX = 'test-zephyr.hex'
 RC_KERNEL_BIN = 'test-zephyr.bin'
 RC_GDB = 'test-none-gdb'
 RC_OPENOCD = 'test-openocd'
-RC_OPENOCD_SEARCH = '/test/openocd/search'
+RC_OPENOCD_SEARCH = ['/test/openocd/search']
 
 
 @pytest.fixture
```

nandojve commented:
Hi @mbolivar-nordic ,

Thanks for bring this to my eyes.

nandojve commented:
Hi @mbolivar-nordic ,

Did you have time to re check after newer changes?

mbolivar-nordic commented:
> I think some unification is needed to make sure this works properly for all boards that support this runner. [...] I also don't think it's safe in general to do a whitespace split (cfg.openocd_search.split()) to get the search path. 

Sorry I was unclear. These were two separate, unrelated comments:

1. additional unification is needed to keep nRF boards working
2. openocd_search should be a list, not a whitespace separated string, if it's going to contain multiple elements

The patch I posted was for 2., not 1. So 1. is still not done. If you try this on an nRF board, you'll get this now:

```
/home/mbolivar/zp/zephyr/boards/common/openocd.zephyr.cfg:4: Error: can't read "_TARGETNAME": no such variable
```

> doesn't play well with boards/common/openocd-nrf5.board.cmake

If you take a look at this file you can see that it's responsible for including the right interface and source files for the nRF dev board, before doing `$_TARGETNAME configure -rtos auto`. However this PR is causing openocd.zephyr.cfg to be loaded before _TARGETNAME is set, so it's breaking those boards.

nandojve commented:
Hi @mbolivar-nordic ,

Ok, now it is clear the whole thing, and, I've been thinking that `openocd.zephyr.cfg` is not necessary anymore.

nandojve committed:
scripts: runners: openocd: Enable thread awareness

Zephyr thread awareness is available for openocd but boards don't
have debuggers configuration.  This configure OpenOCD runner
automatically to complete configuration.

User still require enable CONFIG_DEBUG_THREAD_INFO=y to visualize
thread debug information.

Signed-off-by: Gerson Fernando Budke <nandojve@gmail.com>

mbolivar-nordic reviewed (approved):
Looks good, thanks for testing!

erwango reviewed (approved):
Tested on disco_l475_iot1.
Nice work, thanks!

One suggestion is to add `CONFIG_DEBUG_THREAD_INFO=y` to `samples/basic/threads/prj.conf` as a demonstrator (maybe adding a comment stating this is only valid for openocd users)
